drivers/ethernet/dwc_xgmac/eth_dwc_xgmac.c

```diff
@@ -0,0 +1,1943 @@
+
+				if (!(rx_desc_data.rdes3 & XGMAC_RDES3_ES)) {
+					desc_data_len =
+						(rx_desc_data.rdes3 & XGMAC_RDES3_PL) %
+						(CONFIG_NET_BUF_DATA_SIZE * RX_FRAGS_PER_DESC);
+
+					if (desc_data_len > CONFIG_NET_BUF_DATA_SIZE) {
+						add_frags_to_pkt(
+							rx_desc_meta->rx_pkt, frag1,
+							CONFIG_NET_BUF_DATA_SIZE, frag2,
+							(desc_data_len - CONFIG_NET_BUF_DATA_SIZE));
+					} else {
+						add_frags_to_pkt(rx_desc_meta->rx_pkt, frag1,
+								 desc_data_len, frag2, 0u);
+					}
+					/* Full packet received, submit to net sub system for
+					 * further processing
+					 */
+					err = net_recv_data(data->iface, rx_desc_meta->rx_pkt);
+					if (err) {
+						UPDATE_ETH_STATS_RX_ERROR_PKT_CNT(data, 1u);
+						net_pkt_unref(rx_desc_meta->rx_pkt);
+						LOG_DBG("%s: received packet dropped %d", dev->name,
+							err);
+					} else {
+						LOG_DBG("%s: received a packet", dev->name);
+						UPDATE_ETH_STATS_RX_BYTE_CNT(
+							data,
+							net_pkt_get_len(rx_desc_meta->rx_pkt));
+					}
+				} else {
+					LOG_ERR("%s: rx packet error", dev->name);
+					UPDATE_ETH_STATS_RX_ERROR_PKT_CNT(data, 1u);
+					net_pkt_unref(rx_desc_meta->rx_pkt);
+				}
+				rx_desc_meta->rx_pkt = (struct net_pkt *)NULL;
+			} else {
+				add_frags_to_pkt(rx_desc_meta->rx_pkt, frag1,
+						 CONFIG_NET_BUF_DATA_SIZE, frag2,
+						 CONFIG_NET_BUF_DATA_SIZE);
+			}
+		} else {
+			LOG_ERR("%s: Received a fragment with no FD fragmanet received in the "
+				"sequence",
+				dev->name);
+		}
+		rx_desc_meta->next_to_read =
+			((rx_desc_meta->next_to_read + 1) % dma_chnl_cfg->rdrl);
+		rx_desc = (struct xgmac_dma_rx_desc *)(fisrt_rx_desc + rx_desc_meta->next_to_read);
+		arch_dcache_invd_range(rx_desc, sizeof(rx_desc));
+		rx_desc_data = *(rx_desc);
+	}
+}
+
+static void eth_dwc_xgmac_tx_irq_work(const struct device *dev, uint32_t dma_chnl)
+{
+	struct eth_dwc_xgmac_dev_data *const data = (struct eth_dwc_xgmac_dev_data *)dev->data;
+	const struct eth_dwc_xgmac_config *const config =
+		(struct eth_dwc_xgmac_config *)dev->config;
+	struct xgmac_dma_chnl_config *dma_chnl_cfg =
+		(struct xgmac_dma_chnl_config *)&config->dma_chnl_cfg;
+	struct xgmac_dma_tx_desc_meta *tx_desc_meta =
+		(struct xgmac_dma_tx_desc_meta *)&data->tx_desc_meta[dma_chnl];
+	struct xgmac_dma_tx_desc *fisrt_tx_desc =
+		(struct xgmac_dma_tx_desc *)(data->dma_tx_desc + (dma_chnl * dma_chnl_cfg->tdrl));
+	struct xgmac_dma_tx_desc *tx_desc;
+	uint16_t desc_idx;
+	struct net_pkt *pkt;
+
+	desc_idx =
+		((tx_desc_meta->next_to_use + k_sem_count_get(&tx_desc_meta->free_tx_descs_sem)) %
+		 dma_chnl_cfg->tdrl);
+	for (; desc_idx != tx_desc_meta->next_to_use;
+	     desc_idx = ((desc_idx + 1) % dma_chnl_cfg->tdrl)) {
+		tx_desc = (struct xgmac_dma_tx_desc *)(fisrt_tx_desc + desc_idx);
+		arch_dcache_invd_range(tx_desc, sizeof(tx_desc));
+		if (!(tx_desc->tdes3 & XGMAC_TDES3_OWN)) {
+			if (tx_desc->tdes3 & XGMAC_TDES3_LD) {
+				pkt = (struct net_pkt *)(*(
+					data->tx_pkts +
```

Comment:
it's becomes quickly unreadable, perhaps a quick inline function would be better

Reply:
Added an inline function.